### PR TITLE
[#29] Add proper documentation

### PR DESCRIPTION
**Pull request by PruStephan**
## Description

Add documentation for all necessary entrypoints and datatypes
## Related issue(s)

Resolves first part of #29

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: src/Lorentz/Contracts/BaseDAO.hs [hunk 1]
```diff
@@ -19,21 +19,33 @@ import Lorentz
...
 import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
 
+-- | Common marker for permanent entrypoints.
+-- Can be used when parameter for permanent entrypoints is flat, i.e. does not
+-- have nested subparameters with multiple entrypoints.
+data PermanentEntrypointsKind
+instance DocItem (DEntrypoint PermanentEntrypointsKind) where
+  docItemPos = 1030
+  docItemSectionName = Just "Permanent entrypoints"
+  docItemToMarkdown = diEntrypointToMarkdown
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO.hs:33]: Probably you forgot to rename it? In our case, we need to use `PlainEntrypointsKind` for our top-level `entryCase` and a custom datatype for `entryCase` related to FA2 entrypoints.

Also, make sure that `docItemPos` is quite unique.

File: src/Lorentz/Contracts/BaseDAO.hs [hunk 2]
```diff
@@ -47,8 +59,10 @@ baseDaoContract config@Config{..} = defaultContract $ contractName cDaoName $ do
     , #cFlush /-> flush config
     )
 
-fa2Handler :: (IsoValue pm) => Entrypoint FA2.Parameter (BaseDAO.Storage pm)
-fa2Handler =
+fa2Handler :: forall pm. (IsoValue pm, TypeHasDoc pm) => Entrypoint FA2.Parameter (BaseDAO.Storage pm)
+fa2Handler = do
+  docStorage @(BaseDAO.Storage pm)
+  doc $ DDescription callFA2Doc
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO.hs:65]: This should not be here, you should put it to `instance DocItem (DEntrypoint ...)`, otherwise it will appear dangling in the documentation.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+
+flushDoc :: Markdown
+flushDoc = [md|
+  Finish voting process on all proposals where the voting period is over.
+  Returns an ammount to the proposer, determined by the result of voting.
+  There is a possibility of some tokens beign lost due to admistrator perform
+  of burn or transfer operation.
+  If proposal is accepted, the dicision lambda is called.
+  |]
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:137]: :+1:

File: src/Lorentz/Contracts/BaseDAO.hs [hunk 4]
```diff
@@ -56,7 +68,9 @@ baseDaoContract config@Config{..} = defaultContract $ contractName cDaoName $ do
 fa2Handler
   :: (IsoValue pm, HasFuncContext s (BaseDAO.Storage pm))
   => Entrypoint' FA2.Parameter (BaseDAO.Storage pm) s
-fa2Handler =
+fa2Handler = do
+  docStorage @(BaseDAO.Storage pm)
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO.hs:72]: Why is storage description here?

(not sure if @Martoon-00's comment applies to this line as well)

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+callFA2Doc = [md|
+  Entrypoint to be called if you want to use one of FA2 entrypoints.
+  |]
+
+transferDoc :: Markdown
+transferDoc = [md|
+  Transfer tokens between a given account and each account from the given list.
+
+  It serves multiple purposes:
+  * If transaction `"from"` address equals to the admin address presented in storage, than  it is allowed for
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:42]:
```suggestion
  * If transaction `"from"` address equals to the admin address presented in storage, then it is allowed for
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  any address and both `frozen` and `unfrozen` tokens
+  * Otherwise, it is allowed to transfer money only to the accounts presented as `"from"` operators in storage.
+  It is also permited to send frozen tokens in this case.
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:45]:
```suggestion
  It is also prohibited to send frozen tokens in this case.
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  |]
+
+transferDoc :: Markdown
+transferDoc = [md|
+  Transfer tokens between a given account and each account from the given list.
+
+  It serves multiple purposes:
+  * If transaction `"from"` address equals to the admin address presented in storage, than  it is allowed for
+  any address and both `frozen` and `unfrozen` tokens
+  * Otherwise, it is allowed to transfer money only to the accounts presented as `"from"` operators in storage.
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:44]: I don't understand this sentence. It's allowed to transfer money to any account. The restriction applies to `from` addresses: each address must be `SENDER` or have `SENDER` as an operator.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+  Confirms migration of a contract to the sender address.
+  After a successful call the contract will be set to migrated state where no operations are possible.
+  |]
+
+proposeDoc :: Markdown
+proposeDoc = [md|
+  Saves the proposal with specific id and freezes the ammount of sender tokens equal to the given ammount.
+  Sender must have enough unfrozen tokens.
+  The sender's ammount of frozen tokens is increased by proposal ammount. And the amount of unfrozen one
+  decreased by the same value.
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:103]:
```suggestion
  is decreased by the same value.
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  |]
+
+
+flushDoc :: Markdown
+flushDoc = [md|
+  Finish voting process on all proposals where the voting period is over.
+  Returns an ammount to the proposer, determined by the result of voting.
+  There is a possibility of some tokens beign lost due to admistrator perform
+  of burn or transfer operation.
+  If proposal is accepted, the dicision lambda is called.
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:134]:
```suggestion
  If proposal is accepted, the decision lambda is called.
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  Can be performed only by administrator.
+  It affects all ongoing proposals and all created afterwards.
+  |]
+
+
+flushDoc :: Markdown
+flushDoc = [md|
+  Finish voting process on all proposals where the voting period is over.
+  Returns an ammount to the proposer, determined by the result of voting.
+  There is a possibility of some tokens beign lost due to admistrator perform
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:132]:
```suggestion
  There is a possibility of some tokens being lost due to administrator perform
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  Provides the amount of tokens of the given address.
+  Can be performed only by administrator.
+  |]
+
+transferContractTokensDoc :: Markdown
+transferContractTokensDoc = [md|
+  This entrypoint can be used by the administrator
+  to transfer tokens owned (or operated) by this contract in another FA2 contract.
+  Can be performed only by administrator.
+  All tokens passed as value must be unfrozen.
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:154]: What does this sentence mean? We transfer tokens in another contract, we don't know if there is any notion of "frozen"/"unfrozen" tokens at all.

File: src/Lorentz/Contracts/BaseDAO.hs (hunk 1 above)
Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO.hs:35]: I don't understand you. Do I need to create here a `PermanentEntrypointsKind` datatype and use it only in fa2Handler entryCase?

> Also, make sure that docItemPos is quite unique.
Is it calculated somehow, or just a random number?

File: src/Lorentz/Contracts/BaseDAO/Types.hs [hunk 12]
```diff
@@ -103,7 +103,7 @@ data Config proposalMetadata = Config
 defaultConfig :: Config pm
 defaultConfig = Config
   { cDaoName = "BaseDAO"
-  , cDaoDescription = [md|"BaseDAO description"|]
+  , cDaoDescription = [md|BaseDAO description.|]
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Types.hs:106]: I think it should be something more descriptive, like, I don't know, at least "An example of a very simple DAO contract without any custom checks, extra data and decision lambda".

File: src/Lorentz/Contracts/BaseDAO.hs (hunk 2 above)
Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO.hs:73]: Can you explain?

**Review by gromakovsky — CHANGES_REQUESTED** [review]

File: src/Lorentz/Contracts/BaseDAO.hs (hunk 1 above)
Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO.hs:35]: You need to create `FA2EntrypointsKind`, and use it only in fa2Handler entryCase, right.

> Is it calculated somehow, or just a random number?
It specifies global order on doc items in the rendered documentation. Just assign some similar number, like `1035`

Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO.hs:35]: > Also, make sure that docItemPos is quite unique. 

Is it somehow calculated, or just a random number?
Also, can you explain further about `PermanetnEtrypointsKind` and the next comment

File: src/Lorentz/Contracts/BaseDAO.hs (hunk 2 above)
Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO.hs:73]: Well, speaking short, there should be no descriptions outside of atomic entrypoints (and fa2Handler is not an atomic entrypoints, it is a group of entrypoints). Put `callFA2Doc` into `docItemToMarkdown` of `instance DocItem (DEntrypoint FA2EntrypointKind)` instead.

The problem here is that, as you can see in the resulting documentation, we have a section with all top-level entrypoints, and a separate section with FA2 entrypoints. Unlike in code, in the resulting document FA2 entrypoints are not included into `fa2Handler` top-level entrypoint description because markdown does not work well with too deeply nested pieces.

Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO.hs]: So I have updated instance of `FA2EntrypointKind` with callFA2Doc, but it fails the test, saying that this entrypoint is not documented:   `Entrypoint 'Call_FA2' does not contain any description.`

Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO.hs]: Oh hmm, maybe I was wrong :thinking: 

Sorry for this. Could you when check that if you add documentation here as well then the result looks ok, and we will decide then what to do?

Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO.hs:69]: I added callFA2Doc into fa2Handler and it seems to work properly now

Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO.hs:69]: Nice :+1: 

Sorry again for the confusion, I really forgot how to work with this autodoc.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+   , tokenAddressDoc
+   ) where
+
+import Lorentz
+
+import Util.Markdown
+
+introductoryDoc :: Markdown
+introductoryDoc = [md|
+  It contains standart FA2 entrypoints, plus some extra ones including proposal and
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:36]:
```suggestion
  It contains standard FA2 entrypoints, plus some extra ones including proposal and
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+tokenMetadataRegistryDoc :: Markdown
+tokenMetadataRegistryDoc = [md|
+  Returns contract address that holds token metadata.
+  Token metadata MUST contain the DAO metadata.
+  |]
+
+updateOperatorsDoc :: Markdown
+updateOperatorsDoc = [md|
+  Updates Storage opearators. There are 2 different opportunities:
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:70]: Operators of storage? They don't operate storage :confused: 

```suggestion
  Updates operators. There are 2 different opportunities:
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs (hunk 11 above)
Comment (gromakovsky) [reply src/Lorentz/Contracts/BaseDAO/Doc.hs:162]: Bump. I guess you have taken this statement from `specification.md`, but it was copy-pasted there by mistake, it's wrong and was removed in one PR (which might be not merged yet).

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+  This entrypoint can be used by the administrator
+  to transfer tokens owned (or operated) by this contract in another FA2 contract.
+  Can be performed only by administrator.
+  All tokens passed as value must be unfrozen.
+  |]
+
+tokenAddressDoc :: Markdown
+tokenAddressDoc = [md|
+  Returns the address of the associated FA2 contract.
+  Since FA2 logic is embedded into this contract, this entrypoint always return SELF.
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:168]:
```suggestion
  Since FA2 logic is embedded into this contract, this entrypoint always returns SELF.
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+  |]
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:169]: Github says "No newline at end of file" which is probably bad.

**Review by gromakovsky — CHANGES_REQUESTED** [review]

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+tokenMetadataRegistryDoc :: Markdown
+tokenMetadataRegistryDoc = [md|
+  Returns contract address that holds token metadata.
+  Token metadata MUST contain the DAO metadata.
+  |]
+
+updateOperatorsDoc :: Markdown
+updateOperatorsDoc = [md|
+  Updates opearators. There are 2 different opportunities:
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:70]:
```suggestion
  Updates operators. There are 2 different opportunities:
```

File: test/Test/BaseDAO/Doc.hs [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+import Prelude
+import Test.Tasty (TestTree, testGroup)
+
+import Lorentz (buildLorentzDoc, cCode)
+import Lorentz.Test
+import Lorentz.Contracts.BaseDAO (baseDaoContract, defaultConfig)
+
+test_baseDAO_documentation :: TestTree
+test_baseDAO_documentation = testGroup "baseDAO documentation" $
+  runDocTests testLorentzDoc $ buildLorentzDoc $ cCode $ baseDaoContract (defaultConfig @())
```
Comment (gromakovsky) [inline test/Test/BaseDAO/Doc.hs:17]: Recently `GameDAO` was merged (see #41), we should test its documentation as well.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+  |]
+
+transferDoc :: Markdown
+transferDoc = [md|
+  Transfer tokens between a given account and each account from the given list.
+
+  It serves multiple purposes:
+  * If transaction `"from"` address equals to the admin address presented in storage, then  it is allowed for
+  any address and both `frozen` and `unfrozen` tokens
+  * Otherwise, it is allowed to transfer money only if from address equals to sender address or have sender as an operator
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:52]:
```suggestion
  * Otherwise, it is allowed to transfer money only if `from` address equals to sender address or have sender as an operator.
```

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+balanceOfDoc :: Markdown
+balanceOfDoc = [md|
+  Returns the balance of specified address in ledger.
+  The entrypoint supports both frozen and unfrozen tokens.
+  |]
+
+tokenMetadataRegistryDoc :: Markdown
+tokenMetadataRegistryDoc = [md|
+  Returns contract address that holds token metadata.
+  Token metadata MUST contain the DAO metadata.
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:65]: I think it is better to replace `must` with `will`, since unlike in the specification, here we describe our contract's guarantees.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+  |]
+
+updateOperatorsDoc :: Markdown
+updateOperatorsDoc = [md|
+  Updates opearators. There are 2 different opportunities:
+
+  * Add operator - updates operators with new pair of (operator, owner).
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:72]:
```suggestion
  * Add operator - updates operators with new pair of `(operator, owner)`.
```

Better to enclose all expressions into ticks for clarity (here and below)

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+  If this pair already exists, this entrypoint does nothing
+  * Remove operator - updates operators, removing the existing pair of (operator, owner).
+  If there is no such key, this entrypoint does nothing.
+
+  All tokens passed to this entrypoint must be unfrozen.
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:77]: This sounds like in this entrypoint users can pass tokens. Maybe rephrase somehow like

> This can be applied only to unfrozen tokens

?

File: src/Lorentz/Contracts/BaseDAO/Doc.hs (hunk 21 above)
Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO/Doc.hs:72]: Also please mention what the operator is. Probably providing a link to the [TZIP](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-12/tzip-12.md#operators) should be enough.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+  This entrypoint can be used by the administrator
+  to transfer tokens owned (or operated) by this contract in another FA2 contract.
+  Can be performed only by administrator.
+  |]
+
+tokenAddressDoc :: Markdown
+tokenAddressDoc = [md|
+  Returns the address of the associated FA2 contract.
+  Since FA2 logic is embedded into this contract, this entrypoint always returns SELF.
+  |]
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:168]: In one of the projects, we had a mandatory rule, that if someone writes documentation he has to spellcheck it.

So could you pass the entire text into some spellchecker and ensure that checks are passed? There is https://grammarly.com, and even grammarly plugin to VSCode, or Idea (though not sure whether does it check grammar) - whatever seems more convenient.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+burnDoc :: Markdown
+burnDoc = [md|
+  Reduces the amount of tokens of the given address. Can be performed only by administrator
+  and only if the given address have enought tokens to burn.
+  |]
+
+mintDoc :: Markdown
+mintDoc = [md|
+  Provides the amount of tokens of the given address.
+  Can be performed only by administrator.
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:154]: Predicates like this are better put in code.

Could you remove this sentence everywhere and instead put `doc $ DRequireRole "admin"` into `authorizeAdmin`'s definition? If everything is done correctly, in all the respective places there should be a note about admin to be the only valid caller.

`DRequireRole` is defined in [morley-ledgers](https://gitlab.com/morley-framework/morley-ledgers/blob/master/code/morley-ledgers/src/Lorentz/Contracts/ManagedLedger/Doc.hs#L136) and probably can be just imported.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+  |]
+
+transferContractTokensDoc :: Markdown
+transferContractTokensDoc = [md|
+  This entrypoint can be used by the administrator
+  to transfer tokens owned (or operated) by this contract in another FA2 contract.
+  Can be performed only by administrator.
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:161]: It's important to mention that this entrypoint, unlike others, can be used after the contract is migrated.

File: src/Lorentz/Contracts/BaseDAO/Doc.hs [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+acceptOwnershipDoc = [md|
+  Accepts the administrator privelege.
+  Only works when sender was asked to become an admin and only if it was asked by current admin.
+  |]
+
+migrateDoc :: Markdown
+migrateDoc = [md|
+  Asks an address to migrate the contract to it.
+  Can be performed only by current administrator.
+  The contract is not considered migrated, until it recieves confirm_migration call.
```
Comment (Martoon-00) [inline src/Lorentz/Contracts/BaseDAO/Doc.hs:97]: Nice, this looks more clear than in the spec :+1:

**Review by Martoon-00 — COMMENTED** [review]
New pack of comments :nabros:

File: src/Lorentz/Contracts/BaseDAO/Doc.hs (hunk 23 above)
Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO/Doc.hs:168]: >  Idea has build-in spellchecker 

It doesn't check Markdown for some reason

Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO/Doc.hs:168]: Oh mda, good grief indeed

File: src/Lorentz/Contracts/BaseDAO/Doc.hs (hunk 24 above)
Comment (PruStephan) [reply src/Lorentz/Contracts/BaseDAO/Doc.hs:154]: Actually it is already there ...

File: test/Test/BaseDAO/Doc.hs (hunk 18 above)
Comment (PruStephan) [reply test/Test/BaseDAO/Doc.hs:17]: Am I understanding it right, that here only `gameDaoProposalCheck`, `gameDaoRejectedProposalReturnValue`, `decisionLambda` and `gameDaoContract` should be commented?

Comment (gromakovsky) [reply test/Test/BaseDAO/Doc.hs:17]: How do you want to comment them? If you mean haddock comments, they are already commented.

You need to add `runDocTests testLorentzDoc $ buildLorentzDoc $ cCode gameDaoContract` somewhere here. Maybe this test will pass immediately, but if it fails, you'll need to fix it.

**Comment by gromakovsky** [comment]
Also I recommend rebasing on latest master ASAP to make tests faster.

**Comment by PruStephan** [comment]
Only need to squash all commits, everything else is ready

File: src/Lorentz/Contracts/BaseDAO/Doc.hs (hunk 24 above)
Comment (Martoon-00) [reply src/Lorentz/Contracts/BaseDAO/Doc.hs]: Oh really, memno.

Okay. That was really good that documentation mentioned about administrator explicitly, it's just that we can do this in a smarter way, fonrtunately.

**Review by Martoon-00 — APPROVED** [review]

**Comment by Martoon-00** [comment]
Thanks for resolving my comments, looks pretty good.

**Review by gromakovsky — APPROVED** [review]
Alright, please squash commits and proceed to the merge.
